### ISSUE TITLE
Universe polymorphism seems to generalize over useless internal universes

### ISSUE THREAD

**Issue by herbelin** [coq/coq#13286]
#### Description of the problem

While experimenting with unification (#13285), I discovered that a polymorphic inductive definition may contain a generalization over universes internal to the type-checking of the inductive definition. 

Here is a short example (extracted from [bug_5208.v] (https://github.com/coq/coq/blob/master/test-suite/bugs/closed/bug_5208.v)):
```
Section poly.
Universe U.
Axiom val : Type@{U}.
Inductive fields : Type :=
| branch : option Type@{U} -> fields.
Inductive member : fields -> Type :=
| pm : member (branch (Some val)).
Set Printing Universes.
Print member.
(*
Inductive member : fields -> Prop :=
    pm : member (branch (Some val))
(* {member.u0} |= U < option.u0
                  U = member.u0 *)
*)
```
This looks like some pruning missing. Is that possible?

In any case, after an effort for postponing the `evar <= sort` constraints in #13285, I get:
```
Inductive member : fields -> Prop :=
    pm : member (branch (Some val))
(* {} |= U < option.u0 *)
 ```
which seems to be the expected result.

#### Coq Version

From 8.5 to 8.12 (but the printing of universes greatly improved in the way!!)

**Comment by SkySkimmer** [comment]
With Printing All we see
~~~coq
Inductive member : forall _ : fields, Prop :=
    pm : member (branch (@Some Type@{member.u0} val))
(* {member.u0} |= U < option.u0
                  U = member.u0 *)
~~~
so it seems like a minimization issue rather than a universe restriction issue.

**Comment by herbelin** [comment]
You're right, my analysis is wrong, it is not a pruning issue but rather equations `u = v` involving generalized variables which are not inlined. Let me rephrase the whole comment and change the title (old comment accessible in the modified menu).

**Comment by herbelin** [comment]
Another example illustrating how two similar inference problems lead to different generalizations (where I used symbolic names instead of numbers to report the errors and instantiations):
```
Universe u.
Parameter A : Type@{u}.
     Inductive T1@{i k} : Type@{u} -> Type@{i} := C1 : @id _ (*j*)  (T1 (@id Type@{k} A)) -> T1 A. (*j:=i*)
Fail Inductive T2@{i}   : Type@{u} -> Type@{i} := C2 : @id _ (*j*)  (T2 (@id _ (*k*)  A)) -> T2 A. (*k unbound*)
Fail Inductive T3@{i j} : Type@{u} -> Type@{i} := C3 : @id Type@{j} (T3 (@id _ (*k*)  A)) -> T3 A. (*k unbound*)
```
Debugging shows that in the last two cases we have the equations `k = u` but the equation is not used for inlining `k`.

Is there an issue with `restrict_universe_context`? Or should something be done elsewhere?

**Comment by SkySkimmer** [comment]
>Is there an issue with restrict_universe_context? Or should something be done elsewhere?

As I said this is not a restrict issue.

**Comment by herbelin** [comment]
> As I said this is not a restrict issue.

Do you know what to do?
